Add preliminary support for borrow accessors

meg-gupta opened the pull request:
Introduces a new experimental flag, AST, SILGen and limited SIL support. 

Adds support for writing borrow accessors that return projections of `self` representable by a new `ResultConvention::Guaranteed` or  `ResultConvention::GuaranteedAddress`.

There are still several unimplemented parts. The main ones are:

- mutate accessor support in SIL and lower (`InOutType` is banned)
- borrow and mutate accessors in address-only types with  multiple returns (address phis are banned in SIL)
- borrow accessors in loadable ~Copyable types (SILGen of borrow accessors clashes with late resolution of move-only types)

meg-gupta committed:
Add a new experimental feature for borrow and mutate accessors

meg-gupta committed:
Parser support for borrow and mutate accessors

meg-gupta committed:
Add mangling support for borrow/mutate accessors

meg-gupta committed:
Avoid synthesizing read when borrow is present and modify when mutate is present

meg-gupta committed:
Introduce borrow and mutate as new accessor kinds

And handle them in various covered switches

meg-gupta committed:
Sema support for borrow and mutate accessors

meg-gupta committed:
Add tests for borrow and mutate accessors

meg-gupta committed:
[NFC] Add some utilities in SILDeclRef

meg-gupta committed:
Introduce ResultConvention::Guaranteed and  ResultConvention::GuaranteedAddress

ResultConvention::Guaranteed will be used by borrow accessors when the storage type can be returned by value.

ResultConvention::GuaranteedAddress will be used by mutate accessors and borrow accessors when the storage type
cannot be returned by value.

meg-gupta committed:
Update SILFunctionType for borrow accessors

meg-gupta committed:
Update DeclContext::getInnermostPropertyAccessorContext for accessors within SubscriptDecl

meg-gupta committed:
[NFC] Move StorageRefResult to a header

meg-gupta committed:
Updates to MoveOnlyChecker

meg-gupta committed:
Updates to MemAccessUtils to handle borrowed address returns

meg-gupta committed:
Enable SIL parsing of borrow and mutate accessors

meg-gupta committed:
Introduce ResultConvention::Guaranteed and  ResultConvention::GuaranteedAddress in SwiftCompilerSources

meg-gupta committed:
Update SIL verification for borrow accessors

meg-gupta committed:
Update OperandOwnership of ReturnInst

meg-gupta committed:
SILGen support for borrow accessors on Copyable types and address-only ~Copyable types

meg-gupta committed:
Update borrowed from for guaranteed results of borrow accessors

meg-gupta committed:
Add some optimizer tests

meg-gupta committed:
Update SideEffects for borrow accessors returning addresses

Add a read effect on the self parameter. Without this, the self parameter can get dead code eliminated by the GenericSpecializer.

meg-gupta committed:
Disable GenericSpecializer for borrow accessors in OSSA temporarily

meg-gupta committed:
[NFC] hasGuaranteedAddressResults -> hasGuaranteedAddressResult and hasGuaranteedResults -> hasGuaranteedResult

meg-gupta committed:
Update hasGuaranteedResult and hasGuaranteedAddressResult

meg-gupta commented:
@swift-ci test

meg-gupta committed:
[NFC] Replace ApplyInst::getSubstCalleeConv()::hasGuaranteedAddressResult() with  ApplyInst::hasGuaranteedAddressResult()

meg-gupta commented:
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/11371

@swift-ci Please test

meg-gupta committed:
[NFC] Update feature flags

meg-gupta committed:
Avoid inserting lexical scope for arguments of borrow accessors while inlining

atrick reviewed (approved):
This all looks great to me. I glossed over most of the SILGen implementation because I'm not much use there.

---
We need verification to ensure that, whenever a borrow accessor returns an address, it's implementation only refers to addressable storage recursively up the the addressable 'self' parameter. I know your diagnostics mostly cover this. I just think we need SIL verification on top of that which can run after any pass.

---
We should test IRGen for large loadable values.

---
It's to be determined whether the SIL should model the borrows result with `mark_dependence` . If the apply can always be viewed as an ownership forwarding operation on the 'self' operand, and the we guarantee that the accessor implementation only allows forwarding (or escaping) operations. Then we don't structly need another SIL dependency.

meg-gupta committed:
Update SwiftCompilerSources' OwnershipLiveness utility for borrow accessors

It uses a check on conformance to ForwardInstruction for walking down guaranteed forwarding uses.
Since apply of borrow accessors cannot be represented as ForwardingInstruction, handle them separately.

Representing apply of borrow accessors for consistent handling in the optimizer is TBD.

meg-gupta committed:
[NFC] Add notes on self argument of borrow accessor being non-lexical

meg-gupta commented:
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/11371

@swift-ci Please test

meg-gupta commented:
Please test with following pull request:
https://github.com/swiftlang/llvm-project/pull/11371

@swift-ci Please test
